File: tutorials/workflow/csharp/resiliency-and-compensation/ResiliencyAndCompensation/ResiliencyAndCompensationWorkflow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Dapr.Workflow;
using ResiliencyAndCompensation.Activities;

namespace ResiliencyAndCompensation;

internal sealed class ResiliencyAndCompensationWorkflow : Workflow<int, int?>
{
    public override async Task<int?> RunAsync(WorkflowContext context, int input)
    {
         // This creates a retry policy that retries activities up to 3 times with an initial delay of 2 seconds.
         var defaultActivityRetryOptions = new WorkflowTaskOptions
         {
            RetryPolicy = new WorkflowRetryPolicy(
                maxNumberOfAttempts: 3,
                firstRetryInterval: TimeSpan.FromSeconds(2)),
         };
        
        var result1 = await context.CallActivityAsync<int>(
            nameof(MinusOne),
            input,
            defaultActivityRetryOptions);

        int? workflowResult = null;
        try
        {
            workflowResult = await context.CallActivityAsync<int>(
                nameof(Division),
                result1,
                defaultActivityRetryOptions);
        }
        catch (WorkflowTaskFailedException ex)
        {
            // Something went wrong in the Division activity which is not recoverable.
            // Perform a compensation action for the MinusOne activity to revert any
            // changes made in that activity.
            if (ex.FailureDetails.IsCausedBy<OverflowException>())
            {
                workflowResult = await context.CallActivityAsync<int>(
                    nameof(PlusOne),
                    result1,
                    defaultActivityRetryOptions);

                // A custom status message can be set on the workflow instance.
                // This can be used to clarify anything that happened during the workflow execution.
                context.SetCustomStatus("Compensated MinusOne activity with PlusOne activity.");
            }
        }

        return workflowResult;
    }
}

File: tutorials/workflow/python/resiliency-and-compensation/resiliency_and_compensation/resiliency_and_compensation_workflow.py
from datetime import timedelta
import dapr.ext.workflow as wf

wf_runtime = wf.WorkflowRuntime()

@wf_runtime.workflow(name='resiliency_and_compensation_workflow')
def resiliency_and_compensation_workflow(ctx: wf.DaprWorkflowContext, wf_input: int):
    default_retry_policy = wf.RetryPolicy(max_number_of_attempts=3, first_retry_interval=timedelta(seconds=2))

    result1 = yield ctx.call_activity(minus_one, input=wf_input, retry_policy=default_retry_policy)
    wf_result = None

    try:
        wf_result = yield ctx.call_activity(division, input=result1, retry_policy=default_retry_policy)
    except Exception as e:
        """
        Something went wrong in the division activity which is not recoverable.
        Perform a compensation action for the minus_one activity to revert any
        changes made in that activity.
        """
        if type(e).__name__ == 'TaskFailedError':
            wf_result = yield ctx.call_activity(plus_one, input=result1)

            """
            A custom status message can be set on the workflow instance.
            This can be used to clarify anything that happened during the workflow execution.
            """
            ctx.set_custom_status("Compensated minus_one activity with plus_one activity.")

    return wf_result

@wf_runtime.activity(name='minus_one')
def minus_one(ctx: wf.WorkflowActivityContext, act_input: int) -> int:
    print(f'minus_one: Received input: {act_input}.', flush=True)
    result = act_input - 1
    return result

@wf_runtime.activity(name='division')
def division(ctx: wf.WorkflowActivityContext, act_input: int) -> int:
    """
    In case the divisor is 0 an exception is thrown.
    This exception is handled in the workflow.
    """
    print(f'division: Received input: {act_input}.', flush=True)
    result = int(100/act_input)
    return result

@wf_runtime.activity(name='plus_one')
def plus_one(ctx: wf.WorkflowActivityContext, act_input: int) -> int:
    print(f'plus_one: Received input: {act_input}.', flush=True)
    result = act_input + 1
    return result